File: deepgoal/core/node.py
from __future__ import annotations

from abc import ABC, abstractmethod
from datetime import datetime

from pydantic import BaseModel, Field

from .types.common import JsonValue
from .types.engine import EngineOptions


class Artifact(BaseModel):
    path: str
    kind: str | None = None
    mime_type: str | None = None
    metadata: dict[str, JsonValue] = Field(default_factory=dict)


class ExecutionError(BaseModel):
    type: str
    message: str
    details: dict[str, JsonValue] = Field(default_factory=dict)
    traceback: str | None = None
    recoverable: bool = False


class ExecutionMetrics(BaseModel):
    started_at: datetime | None = None
    finished_at: datetime | None = None
    duration_seconds: float | None = None


class PipelineInput(BaseModel):
    primary_path: str | None = None
    artifacts: list[Artifact] = Field(default_factory=list)
    metadata: dict[str, JsonValue] = Field(default_factory=dict)

    @classmethod
    def from_path(cls, path: str) -> PipelineInput:
        return cls(primary_path=path)

    def to_output(self) -> PipelineOutput:
        return PipelineOutput(primary_path=self.primary_path, artifacts=list(self.artifacts), metadata=dict(self.metadata))


class PipelineOutput(BaseModel):
    primary_path: str | None = None
    artifacts: list[Artifact] = Field(default_factory=list)
    metadata: dict[str, JsonValue] = Field(default_factory=dict)

    @classmethod
    def from_path(cls, path: str) -> PipelineOutput:
        return cls(primary_path=path)

    def to_input(self) -> PipelineInput:
        return PipelineInput(primary_path=self.primary_path, artifacts=list(self.artifacts), metadata=dict(self.metadata))


class PipelineResult(BaseModel):
    success: bool
    output: PipelineOutput = Field(default_factory=PipelineOutput)
    error: ExecutionError | None = None
    metrics: ExecutionMetrics | None = None

    @classmethod
    def ok(cls, output: PipelineOutput | None = None, metrics: ExecutionMetrics | None = None) -> PipelineResult:
        return cls(success=True, output=output or PipelineOutput(), metrics=metrics)

    @classmethod
    def fail(
        cls,
        error: ExecutionError,
        output: PipelineOutput | None = None,
        metrics: ExecutionMetrics | None = None,
    ) -> PipelineResult:
        return cls(success=False, output=output or PipelineOutput(), error=error, metrics=metrics)


class Node(ABC):
    """Pipeline container with an explicit input/output boundary."""

    @abstractmethod
    async def run(self, input: PipelineInput, options: EngineOptions) -> PipelineResult: ...

File: deepgoal-ts/src/core/node.ts
import type { EngineOptions, JsonObject } from './types/index.js';

export interface Artifact {
  path: string;
  kind?: string | undefined;
  mimeType?: string | undefined;
  metadata: JsonObject;
}

export interface ExecutionError {
  type: string;
  message: string;
  details: JsonObject;
  traceback?: string | undefined;
  recoverable: boolean;
}

export interface ExecutionMetrics {
  startedAt?: Date | undefined;
  finishedAt?: Date | undefined;
  durationSeconds?: number | undefined;
}

export interface PipelineInput {
  primaryPath?: string | undefined;
  artifacts: Artifact[];
  metadata: JsonObject;
}

export interface PipelineOutput {
  primaryPath?: string | undefined;
  artifacts: Artifact[];
  metadata: JsonObject;
}

export interface PipelineResult {
  success: boolean;
  output: PipelineOutput;
  error?: ExecutionError | undefined;
  metrics?: ExecutionMetrics | undefined;
}

export abstract class Node {
  abstract run(input: PipelineInput, options: EngineOptions): Promise<PipelineResult>;
}

export function createArtifact(path: string, options: Partial<Omit<Artifact, 'path'>> = {}): Artifact {
  return {
    path,
    metadata: options.metadata ?? {},
    ...(options.kind === undefined ? {} : { kind: options.kind }),
    ...(options.mimeType === undefined ? {} : { mimeType: options.mimeType }),
  };
}

export function createExecutionError(error: Pick<ExecutionError, 'type' | 'message'> & Partial<ExecutionError>): ExecutionError {
  return {
    type: error.type,
    message: error.message,
    details: error.details ?? {},
    recoverable: error.recoverable ?? false,
    ...(error.traceback === undefined ? {} : { traceback: error.traceback }),
  };
}

export function createExecutionMetrics(metrics: Partial<ExecutionMetrics> = {}): ExecutionMetrics {
  return {
    ...(metrics.startedAt === undefined ? {} : { startedAt: metrics.startedAt }),
    ...(metrics.finishedAt === undefined ? {} : { finishedAt: metrics.finishedAt }),
    ...(metrics.durationSeconds === undefined ? {} : { durationSeconds: metrics.durationSeconds }),
  };
}

export function createPipelineInput(input: Partial<PipelineInput> = {}): PipelineInput {
  return {
    ...(input.primaryPath === undefined ? {} : { primaryPath: input.primaryPath }),
    artifacts: [...(input.artifacts ?? [])],
    metadata: { ...(input.metadata ?? {}) },
  };
}

export function pipelineInputFromPath(path: string): PipelineInput {
  return createPipelineInput({ primaryPath: path });
}

export function createPipelineOutput(output: Partial<PipelineOutput> = {}): PipelineOutput {
  return {
    ...(output.primaryPath === undefined ? {} : { primaryPath: output.primaryPath }),
    artifacts: [...(output.artifacts ?? [])],
    metadata: { ...(output.metadata ?? {}) },
  };
}

export function pipelineOutputFromPath(path: string): PipelineOutput {
  return createPipelineOutput({ primaryPath: path });
}

export function inputToOutput(input: PipelineInput): PipelineOutput {
  return createPipelineOutput({
    primaryPath: input.primaryPath,
    artifacts: input.artifacts,
    metadata: input.metadata,
  });
}

export function outputToStandaloneInput(output: PipelineOutput): PipelineInput {
  return createPipelineInput({
    primaryPath: output.primaryPath,
    artifacts: output.artifacts,
    metadata: output.metadata,
  });
}

export function okResult(output: PipelineOutput = createPipelineOutput(), metrics?: ExecutionMetrics): PipelineResult {
  return {
    success: true,
    output,
    ...(metrics === undefined ? {} : { metrics }),
  };
}

export function failResult(error: ExecutionError, output: PipelineOutput = createPipelineOutput(), metrics?: ExecutionMetrics): PipelineResult {
  return {
    success: false,
    output,
    error,
    ...(metrics === undefined ? {} : { metrics }),
  };
}
